### ISSUE TITLE
`kapp deploy --diff-run` should not create app if it doesnt exist

### ISSUE THREAD

**Issue by cppforlife** [vmware-tanzu/carvel-kapp#23]
this was an intentional behaviour to keep code as easy as possible at the time. may be it's time to reconsider.

**Comment by pivotaljohn** [comment]
As mentioned in [this Slack thread](https://kubernetes.slack.com/archives/CH8KCCKA5/p1613595378324200), this also happens _without_ the `--diff-run` flag.

```console
$ kapp deploy -a ambassador
Target cluster 'https://127.0.0.1:55000' (nodes: kind-control-plane)

kapp: Error: Trying to apply empty set of resources will result in deletion of resources on cluster. Refusing to continue unless --dangerous-allow-empty-list-of-resources is specified.

$ kapp list
Target cluster 'https://127.0.0.1:55000' (nodes: kind-control-plane)

Apps in namespace 'default'

Name        Namespaces  Lcs  Lca
ambassador  -           -    -
```

**Comment by caphrim007** [comment]
In my situation I'm running the command as @pivotaljohn mentioned (without --diff-run).

It was a surprise to me because I ran the deploy to see what it was going to do, then typed "n" to not continue, and then did the `kapp list` (I think because I was checking whether I had previously deployed it) and I saw that empty app.

That left me wondering if I had previously deployed the app, so I listed namespaces and didnt see the one that would have been created. Then I deleted the "empty" app, thought it was "oh i must have just done something long ago in the past that created that empty one". Re-ran the `kapp deploy`, listed apps and saw the empty one again and went "hmm...."

So in my case fixing this is a tidiness-related fix, but it doubles as an ops fix because it won't weird out any future operator if they suddenly see a lot of empty apps.

My 2c

**Comment by joaopapereira** [comment]
I was able to replicate the issue as well,

```
$ kapp deploy -a ambassador
Target cluster 'https://127.0.0.1:51295' (nodes: kind-control-plane)

kapp: Error: Trying to apply empty set of resources will result in deletion of resources on cluster. Refusing to continue unless --dangerous-allow-empty-list-of-resources is specified.

$ kapp deploy -a ambassador1 --diff-run
Target cluster 'https://127.0.0.1:51295' (nodes: kind-control-plane)

kapp: Error: Trying to apply empty set of resources will result in deletion of resources on cluster. Refusing to continue unless --dangerous-allow-empty-list-of-resources is specified.

$ kapp list
Target cluster 'https://127.0.0.1:51295' (nodes: kind-control-plane)

Apps in namespace 'default'

Name                 Namespaces        Lcs   Lca
ambassador           -                 -     -
ambassador1          -                 -     -

Lcs: Last Change Successful
Lca: Last Change Age

2 apps

Succeeded

$ kubectl get configmap
NAME                               DATA   AGE
ambassador                         1      26s
ambassador1                        1       6s
```

Will move this issue to our list to be prioritized.

---

Vote on this request

This is an invitation to the community to vote on issues, to help us prioritize our backlog. Use the "smiley face" up to the right of this comment to vote.

👍 "I would like to see this bug fixed as soon as possible"
👎 "There are more important bugs to focus on right now"

We are also happy to receive and review Pull Requests if you want to help fix this issue.

**Comment by ralbertsasrc** [comment]
A fix would be great!  Is there a work around?

**Comment by cppforlife** [comment]
@ralbertsasrc no workaround for this unfortunately. if you are always certain that there shouldnt be an app, then you can run kapp delete after --diff-run to remove created configmap.

**Comment by renuy** [comment]
We are intentionally saving the label for kapp commands. --diff-run is a special case in which we are not applying anything to the cluster, hence, the label need not be created.
This issue should handle only, --diff-run. deploy failures are behaving as expected.

**Comment by sethiyash** [comment]
We have stopped creating kapp app while doing `kapp deploy` with `--diff-run` flag.  The fix is available from release v0.48.0

**Comment by vixus0** [comment]
Under what conditions does `--diff-run` still attempt to change the kapp configmap? I'm seeing the following behaviour on v0.58.0:

```
Run carvel-dev/setup-action@v1
Installing ytt:latest, kapp:latest
Downloading kapp v0.58.0 from https://github.com/carvel-dev/kapp/releases/download/v0.58.0/kapp-linux-amd64
Downloading ytt v0.45.3 from https://github.com/carvel-dev/ytt/releases/download/v0.45.3/ytt-linux-amd64
✅  Verified checksum: "a6729fb8514f10ab58f9ed3b50cd90ef79bf16d1cb29173baa84e1af0bc5ad4f  ./ytt-linux-amd64"
✅  Verified checksum: "b253ea9cf6add07f9497955147dc12e8612c24c36dc9929c9a4fecdc76752bd3  ./kapp-linux-amd64"
download apps: 2.525s

...

Run kapp deploy --app "svc-kapp" --diff-run -f kapp.yaml
Target cluster 'https://XXXXX.eks.amazonaws.com'
kapp: Error: Updating app: configmaps "svc-kapp" is forbidden:

  User "XXXXXXXX" cannot update resource "configmaps" in API group "" in the namespace "svc"
```
